WallsAndHoles/m2mtilemesher.cpp

```diff
@@ -1,233 +1,360 @@
+                                    (startHeight + endHeight) * 0.5,
+                                    xzTileCenter.y() + away.y());
 
+        QVector2D direction = SideTools::awayDirection(sideDirection);
+        float height = startHeight - endHeight;
+        bool upsideDown = false;
 
+        if (height < 0) {
+            height = -height;
+            upsideDown = true;
+            direction = -direction;
+        }
 
-            float h = sideTopBase - baseHeight;
-            float ty = 1 - .5 * h;
-            SimpleTexturedObject::TriangleTexCoords dirt1 = { {0, ty}, {0.5, ty}, {0.5, 1} };
-            SimpleTexturedObject::TriangleTexCoords dirt2 = { {0, ty}, {0.5, 1}, {0, 1} };
-            QVector<SimpleTexturedObject::TriangleTexCoords> bottomTexCoords = {
-                dirt1, dirt2,
-                dirt1, dirt2,
-                dirt1, dirt2,
-                dirt1, dirt2
-            };
 
+        meshData.addQuad(Quad::makeVerticalQuad(
+                             sideCenter,
+                             direction,
+                             thisTile.thickness(),
+                             height,
+                             thisTile.sideImage(),
+                             thisTile.sideMaterial(),
+                             upsideDown));
+    }
 
-            QVector<SimpleTexturedObject::Triangle> bottomTriangles = {
-                {20, 21, 22},
-                {20, 22, 23},
 
-                {24, 25, 26},
-                {24, 26, 27},
+    if (shouldOutputGround) {
+        // The tile is not full thickness, so this wall is responsible for outputting
+        // a piece of the ground beneath the tile. The specific piece it outputs depends
+        // on its orientation:
+        /*
+          (Looking from above in 3D space)
+
+              ^ +Z
+              |
+          <---
+          +X
+
+           _ _ _ _ _ _
+          |\    N    /|
+          | \ _ _ _ / |
+          |  |thin |  |
+          |W |tile | E|
+          |  |_ _ _|  |
+          | /       \ |
+          |/_ _ S _ _\|
+
+         */
+
+        auto groundImage = thisTile.groundInfo().groundImage;
+        auto groundMaterial = thisTile.groundInfo().groundMaterial;
+
+        meshData.addQuad(SideTools::makeTopQuadFragment(sideDirection,
+                                                        offset,                 // Center of ground tile in XZ coordinates.
+                                                        xzTileCenter,           // Center of tile in XZ coordinates.
+                                                        thisTile.thickness(),
+                                                        groundImage,
+                                                        groundMaterial));
+    }
+}
+
 
-                {28, 29, 30},
-                {28, 30, 31},
 
-                {32, 33, 34},
-                {32, 34, 35}
-            };
+/* BEGIN TileInfo */
 
+M2M::TileInfo::TileInfo()
```

Comment:
Please mimic the order used in the .h with the definitions

Reply:
That order will likely change.